implementations/rust/ockam/ockam_command/src/identity/delete.rs

```diff
@@ -31,26 +37,83 @@ impl DeleteCommand {
     }
 }
 
+pub struct DeleteTui {
+    opts: CommandGlobalOpts,
+    cmd: DeleteCommand,
+}
+
+impl DeleteTui {
+    pub async fn run(opts: CommandGlobalOpts, cmd: DeleteCommand) -> miette::Result<()> {
+        let tui = Self { opts, cmd };
+        tui.delete().await
+    }
+}
+
 async fn run_impl(
     _ctx: Context,
     (opts, cmd): (CommandGlobalOpts, DeleteCommand),
 ) -> miette::Result<()> {
-    let state = opts.state;
-    let idt = state.identities.get(&cmd.name)?;
-    if opts
-        .terminal
-        .confirmed_with_flag_or_prompt(cmd.yes, "Are you sure you want to delete this identity?")?
-    {
+    DeleteTui::run(opts, cmd).await
+}
+
+#[ockam_core::async_trait]
+impl DeleteCommandTui for DeleteTui {
+    const ITEM_NAME: &'static str = "identities";
+    fn cmd_arg_item_name(&self) -> Option<&str> {
+        self.cmd.name.as_deref()
+    }
+
+    fn cmd_arg_delete_all(&self) -> bool {
+        self.cmd.all
+    }
+
+    fn cmd_arg_confirm_deletion(&self) -> bool {
+        self.cmd.yes
+    }
+
+    fn terminal(&self) -> Terminal<TerminalStream<Term>> {
+        self.opts.terminal.clone()
+    }
+
+    async fn get_arg_item_name_or_default(&self) -> miette::Result<String> {
+        Ok(get_identity_name(&self.opts.state, &self.cmd.name))
+    }
+
+    async fn list_items_names(&self) -> miette::Result<Vec<String>> {
+        Ok(self.opts.state.identities.list_items_names()?)
+    }
+
+    async fn delete_single(&self, item_name: &str) -> miette::Result<()> {
+        let state = &self.opts.state;
+        let idt = state.identities.get(item_name)?;
         state.delete_identity(idt)?;
-        opts.terminal
+        self.terminal()
             .stdout()
             .plain(fmt_ok!(
                 "The identity named '{}' has been deleted",
-                &cmd.name
+                item_name
             ))
-            .machine(&cmd.name)
-            .json(serde_json::json!({ "name": &cmd.name }))
+            .machine(item_name)
+            .json(serde_json::json!({ "name": item_name }))
             .write_line()?;
+        Ok(())
+    }
+
+    async fn delete_multiple(&self, selected_items_names: Vec<String>) -> miette::Result<()> {
+        let plain = selected_items_names
+            .iter()
+            .map(|name| {
+                //Should be safe to unwrap since the identity come from the list of existing
+                //identities
+                let idt = self.opts.state.identities.get(name).unwrap();
```

Comment:
Even if it's safe to unwrap, you should be able to handle the error right? The fewer unwraps the better.

Reply:
Got it, the only thing is:
1. should the function return early with a miette error? Since fails to retrieve a name that was given from the list should be a bug theoretically
2. or should simply display the *Failed to delete identity ...*

## 1.
```rust
        let plain = selected_items_names
            .iter()
            .map(|name| {
                let idt = self.opts.state.identities.get(name)?;
                if self.opts.state.delete_identity(idt).is_ok() {
                    Ok(fmt_ok!("Identity '{name}' deleted\n"))
                } else {
                    Ok(fmt_warn!("Failed to delete identity '{name}'\n"))
                }
            })
            .collect::<miette::Result<String>>()?;
        self.terminal().stdout().plain(plain).write_line()?;
```

## 2. 
```rust
        let plain = selected_items_names
            .iter()
            .map(|name| {
                if let Ok(idt) = self.opts.state.identities.get(name) {
                    if self.opts.state.delete_identity(idt).is_ok() {
                        fmt_ok!("Identity '{name}' deleted\n")
                    } else {
                        fmt_warn!("Failed to delete identity '{name}'\n")
                    }
                } else {
                    fmt_warn!("Failed to delete identity '{name}'\n")
                }
            })
            .collect::<String>();
        self.terminal().stdout().plain(plain).write_line()?;

```

Reply:
It's ok to just call `let idt = state.identities.get(item_name)?;`. As you said, we already retrieved that identity and should be there. Of course, the same command could be run in a different process between the two `get` operations, in which case we would just return a `not found` error. So, no need for custom error handling here.